eyelid-match-ops/src/primitives/poly/modular_poly/modulus.rs

```diff
@@ -86,7 +85,28 @@ pub fn mod_poly_ark_ref(dividend: &Poly) -> Poly {
 ///
 /// This uses an [`ark-poly`] library implementation, and entirely replaces the inner polynomial representation.
 #[cfg(inefficient)]
-pub fn mod_poly_ark_mut(dividend: &mut Poly) {
+pub fn mod_poly_ark_mut<const MAX_POLY_DEGREE: usize>(dividend: &mut Poly<MAX_POLY_DEGREE>) {
     let remainder = mod_poly_ark_ref(dividend);
     *dividend = remainder;
 }
+
+/// Constructs and returns a new polynomial modulus used for the polynomial field, `X^[MAX_POLY_DEGREE] + 1`.
+/// This means that `X^[MAX_POLY_DEGREE] = -1`.
+///
+/// This is the canonical but un-reduced form of the modulus, because the reduced form is the zero polynomial.
+///
+/// TODO: work out how to generically make this a lazy static.
```

Comment:
One approach is to replace `const` generic with a trait like this:

```rust
trait PolyConfig {
    fn max_degree() -> usize;

    fn poly_modulus() -> Poly<Self>;

    // … maybe also Coeff type.
}
```

And `FULL_RES_POLY_DEGREE` is replaced with:

```rust
struct FullResConfig;

impl PolyConfig for FullResConfig {
    fn max_degree() -> usize {
        return 2048;
    }
    …
}

Poly<FullResConfig>
```

Reply:
The advantages are
1. Easier to include pre-computations with `lazy_static` because the config is a concrete type.
2. Multiple parameters can be grouped together without increasing the verbosity of generics (only `<C: PolyConfig>`).

Reply:
Sounds good! We don't need it yet, because this function is only used in production in PR #48, and only in the slow path (key generation).

But it would be useful to have it around as a ticket if we need it later.

Reply:
> 1. Easier to include pre-computations with `lazy_static` because the config is a concrete type.

I think we'd still need to declare a `lazy_static` per concrete type, just like we would have to for each const generic value.

But with concrete types we could avoid that using a crate like `generic_singleton`, where [`get_or_init_thread_local!()`](https://docs.rs/generic_singleton/0.5.0/generic_singleton/#thread-local-variant) is likely to be the most performant.

But I'm not sure if `lazy_static` or `generic_singleton` will be faster than freshly allocating and cheaply initialising a temporary. We should benchmark in a separate ticket, if the performance becomes an issue.

For now I went with the simplest code that gives us flexible polynomial sizes.

Reply:
> ```rust
>  // … maybe also Coeff type.
> ```

Yes, I think we'd want `Coeff` in the trait, and it might also help to have named calculated constants for:
* `POLY_MODULUS_DEGREE` (currently called `MAX_POLY_DEGREE`)
* `REDUCED_POLY_MAX_COEFFS = POLY_MODULUS_DEGREE` (currently also called `MAX_POLY_DEGREE`)
* `REDUCED_POLY_MAX_DEGREE = REDUCED_POLY_MAX_DEGREE - 1` (currently in code as `MAX_POLY_DEGREE - 1`)

This could avoid the kinds of confusion you mentioned below. But I don't think it's a blocker for this PR.